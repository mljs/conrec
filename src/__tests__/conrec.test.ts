import fs from 'fs';

import { NumberMatrix } from 'cheminfo-types';
import { convert } from 'jcampconverter';

import { Conrec } from '..';

const data = fs.readFileSync(`${__dirname}/data/zhmbc_0.jdx`, 'utf8');
const parsed = convert(data, { noContour: true }).flatten[0];
const matrix: NumberMatrix = parsed.minMax?.z || [];
describe('conrec basic test', () => {
  it('no result because level too far', () => {
    const conrec = new Conrec(matrix);
    const { contours, timeout } = conrec.drawContour({
      contourDrawer: 'basic',
      levels: [-1000000000, 1000000000],
      timeout: 10000,
    });
    expect(timeout).toBeFalsy();
    expect(contours).toStrictEqual([
      { lines: [], zValue: -1000000000 },
      { lines: [], zValue: 1000000000 },
    ]);
  });

  it('2 specified levels', () => {
    const conrec = new Conrec(matrix);
    const { contours, timeout } = conrec.drawContour({
      contourDrawer: 'basic',
      levels: [-100000, 100000],
      timeout: 10000,
    });
    expect(timeout).toBeFalsy();
    expect(contours).toHaveLength(2);
    expect(contours[0].lines).toHaveLength(36864);
    expect(contours[1].lines).toHaveLength(119720);
  });

  it('no levels', () => {
    const conrec = new Conrec(matrix);
    const start = Date.now();
    const { contours, timeout } = conrec.drawContour({
      contourDrawer: 'basic',
      levels: [],
      timeout: 10000,
    });
    expect(timeout).toBeFalsy();
    expect(contours).toHaveLength(0);
    expect(Date.now() - start).toBeLessThan(5);
  });

  it('auto select levels', () => {
    const conrec = new Conrec(matrix);
    const { contours, timeout } = conrec.drawContour({
      contourDrawer: 'basic',
      nbLevels: 10,
      timeout: 10000,
    });

    expect(timeout).toBeFalsy();
    expect(contours).toHaveLength(10);
    expect(contours[0].lines).toHaveLength(0);
    expect(contours[1].lines).toHaveLength(4984);
    expect(contours[8].lines).toHaveLength(32);
    expect(contours[8].lines).toMatchSnapshot();
    expect(contours[9].lines).toHaveLength(0);
  });

  it('auto select levels with swapAxes', () => {
    const conrec = new Conrec(matrix, { swapAxes: true });
    const { contours, timeout } = conrec.drawContour({
      contourDrawer: 'basic',
      nbLevels: 10,
      timeout: 10000,
    });

    expect(timeout).toBeFalsy();
    expect(contours).toHaveLength(10);
    expect(contours[0].lines).toHaveLength(0);
    expect(contours[1].lines).toHaveLength(4984);
    expect(contours[8].lines).toHaveLength(32);
    expect(contours[8].lines).toMatchSnapshot();
    expect(contours[9].lines).toHaveLength(0);
  });

  it('return available contours within 100ms', () => {
    const conrec = new Conrec(matrix);
    const { contours, timeout } = conrec.drawContour({
      contourDrawer: 'basic',
      nbLevels: 10,
      timeout: 10,
    });
    expect(timeout).toBeTruthy();
    expect(contours).toHaveLength(10);
    expect(contours[0].lines).toHaveLength(0);
    expect(contours[1].lines.length).toBeLessThan(2000);
    expect(contours[8].lines).toHaveLength(0);
    expect(contours[9].lines).toHaveLength(0);
  });
});
