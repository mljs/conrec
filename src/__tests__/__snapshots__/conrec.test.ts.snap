// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conrec basic test auto select levels 1`] = `
[
  167.3376920421198,
  182.66230795788022,
  167,
  182.4704966842604,
  167.82152230971127,
  183,
  167.3376920421198,
  182.66230795788022,
  167,
  183.50353925353926,
  167.32331370726166,
  183.32331370726166,
  167.32331370726166,
  183.32331370726166,
  167.82152230971127,
  183,
  166.68422114608558,
  182.68422114608558,
  166.30239814567176,
  183,
  167,
  182.4704966842604,
  166.68422114608558,
  182.68422114608558,
  166.69396534866442,
  183.3060346513356,
  167,
  183.50353925353926,
  166.30239814567176,
  183,
  166.69396534866442,
  183.3060346513356,
]
`;

exports[`conrec basic test auto select levels with swapAxes 1`] = `
[
  182.66230795788022,
  167.3376920421198,
  182.4704966842604,
  167,
  183,
  167.82152230971127,
  182.66230795788022,
  167.3376920421198,
  183.50353925353926,
  167,
  183.32331370726166,
  167.32331370726166,
  183.32331370726166,
  167.32331370726166,
  183,
  167.82152230971127,
  182.68422114608558,
  166.68422114608558,
  183,
  166.30239814567176,
  182.4704966842604,
  167,
  182.68422114608558,
  166.68422114608558,
  183.3060346513356,
  166.69396534866442,
  183.50353925353926,
  167,
  183,
  166.30239814567176,
  183.3060346513356,
  166.69396534866442,
]
`;
